torch/_lazy/__init__.py

```diff
@@ -20,3 +25,105 @@ def wait_device_ops(devices=None):
+    Since dist._Work doesn't provide a way to set the value in its torch.futures.Future object, we have
+    to rely on this hook to set the right value.
+
+    Args:
+        state (object): Not used at all.
+        bucket: (dist.GradBucket): The bucket of gradients for all_reduce.
+
+    Returns:
+        A Future that contains a LazyTensor.
+    """
+    assert os.environ.get('LTC_TS_CUDA', False)
+
+    buffer = bucket.buffer()
+
+    cuda_buffer = buffer.to(torch.device('cuda', buffer.device.index))
+    torch.distributed.all_reduce(cuda_buffer)
+    buffer.copy_(cuda_buffer)
+
+    fut = torch.futures.Future()
+    fut.set_result(buffer)
+
+    return fut
+
+
+class LazyProcessGroup(dist.ProcessGroup):
+    """
+    A class to build a custom dist.ProcessGroup instance which overwrites
+    the ``dist.ProcessGroup.getBackendName()``, ``dist.ProcessGroup.allgather()``,
+    ``dist.ProcessGroup.allreduce()``, ``dist.ProcessGroup.broadcast()``
+    methods to handle lazy devices. It currently only works for NCCL and
+    CUDA backend devices.
+    Args:
+        store (dist.Store): key/value store accessible to all workers, used
+                            to exchange connection/address information.
+        rank (int): Rank of the current process (it should be a
+                    number between 0 and ``size``-1).
+        size (int): Number of processes participating in the job.
+        timeout (timedelta): Timeout for operations executed against
+                             the process group.
+    """
+    def __init__(self, store: dist.Store, rank: int, size: int, timeout: timedelta):
+        assert os.environ.get('LTC_TS_CUDA', False)
+        dist.ProcessGroup.__init__(self, rank, size)
+        self.nccl_pg = dist.ProcessGroupNCCL(store, rank, size, timeout)
+
+    def getBackendName(self):
+        return "Lazy"
+
+    def allgather(self, output_tensor_lists, input_tensor_list, opts=None):
+        # The device index in output_tensor_lists could be wrong. Let's use the input_tensor_list's.
+        cuda_output_tensor_lists = [
+            [tensor.to(torch.device('cuda', input_tensor_list[0].device.index)) for tensor in tensor_list]
+            for tensor_list in output_tensor_lists
+        ]
+        cuda_input_tensor_list = [tensor.to(torch.device('cuda', tensor.device.index)) for tensor in input_tensor_list]
+        if opts == None:
+            work = self.nccl_pg.allgather(cuda_output_tensor_lists, cuda_input_tensor_list)
+        else:
+            work = self.nccl_pg.allgather(cuda_output_tensor_lists, cuda_input_tensor_list, opts)
+        work.wait()
+        for tensor_list, cuda_tensor_list in zip(output_tensor_lists, cuda_output_tensor_lists):
+            for tensor, cuda_tensor in zip(tensor_list, cuda_tensor_list):
+                tensor.copy_(cuda_tensor)
+        return work
+
+    def allreduce(self, tensor_list, opts=None):
```

Comment:
so.. this 'allreduce' works in some cases (maybe just in test cases?) but the 'comm hook' is the main pathway that actually gets used right?  why not have the comm_hook function call this one? (work obj is an inconvenience?)

Reply:
It works if you directly calls `dist.all_reduce(...)`. We can't let the comm_hook to call this because we can't create a work objc with and set it's Future. The comm_hook is like a tentative workaround for this.